perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_groupNames.java

```diff
@@ -37,25 +40,32 @@
 	private static final String FRIENDLY_NAME = "groupNames";
 	private static final String A_U_V_GROUP_NAMES = AttributesManager.NS_USER_ATTR_VIRT + ":" + FRIENDLY_NAME;
 
-
 	@Override
 	public Attribute getAttributeValue(PerunSessionImpl sess, User user, AttributeDefinition attributeDefinition) throws InternalErrorException {
 		Attribute attribute = new Attribute(attributeDefinition);
 		List<String> groupNames = new ArrayList<>();
+		List<Map<String, Object>> names = sess.getPerunBl().getDatabaseManagerBl().getJdbcPerunTemplate().queryForList("SELECT" +
+			" DISTINCT vos.short_name AS vo_short_name, groups.name AS group_name" +
+			" FROM" +
+			" members" +
+			" JOIN vos ON vos.id = members.vo_id" +
+			" JOIN groups_members ON groups_members.member_id = members.id" +
+			" JOIN groups ON groups_members.group_id = groups.id" +
+			" WHERE" +
```

Comment:
``WHERE`` clause is resolved after all joins on tables are performed. This can cause performance issue, since DB will first join whole (possibly large) tables and then perform limitation, discarding most data. While today DBs are optimized for this, it's still better to include limitation as a part of ``JOIN`` caluse.

You can put ``members.user_id = ?`` condition as a part of join with ``vos`` table using ``AND`` operator. Like this: ``JOIN vos ON vos.id = members.vo_id AND members.user_id = ?``.

Then do the same with ``groups_members.source_group_status = ?`` and join on ``groups_members`` table.

After this, fix order of params passed to queryForList(), since they will be switched by above change.